contracts/liquidity_hub/pool-network/terraswap_pair/src/helpers.rs

```diff
@@ -5,44 +5,345 @@ use cosmwasm_schema::cw_serde;
 use cosmwasm_std::{Decimal, Decimal256, StdError, StdResult, Storage, Uint128, Uint256};
 use cw_storage_plus::Item;
 
-use terraswap::asset::{Asset, AssetInfo};
+use terraswap::asset::{Asset, AssetInfo, PairType};
 use terraswap::pair::PoolFee;
 
 use crate::error::ContractError;
+use crate::math::Decimal256Helper;
+
+/// The amount of iterations to perform when calculating the Newton-Raphson approximation.
+const NEWTON_ITERATIONS: u64 = 32;
+
+// the number of pools in the pair
+const N_COINS: Uint256 = Uint256::from_u128(2);
+
+fn calculate_stableswap_d(
+    offer_pool: Decimal256,
+    ask_pool: Decimal256,
+    amp: &u64,
+    precision: u8,
+) -> Result<Decimal256, ContractError> {
+    let n_coins = Decimal256::from_ratio(N_COINS, Uint256::from_u128(1));
+
+    let sum_pools = offer_pool.checked_add(ask_pool)?;
+    if sum_pools.is_zero() {
+        // there was nothing to swap, return `0`.
+        return Ok(Decimal256::zero());
```

Comment:
this case here is not being tested

Reply:
Good catch. Definitely should test this one.